libvisual-plugins/plugins/actor/infinite/display.c

```diff
@@ -162,15 +162,27 @@ void _inf_compute_surface(InfinitePrivate *priv, t_interpol* vector_field)
 		for (j=0;j<priv->plugheight;j++) {
 			for (i=0;i<priv->plugwidth;i++) {
 				interpol = &vector_field[add_dest];
-				add_src = (interpol->coord & 0xffff) * priv->plugwidth + (interpol->coord >> 16);
+				const uint16_t y = interpol->coord & 0xffff;
+				const uint16_t x = interpol->coord >> 16;
+				add_src = y * priv->plugwidth + x;
 				ptr_pix = priv->surface1 + add_src;;
 
-				/* FIXME it does buffer overread here now and then */
-
-				color= (*(ptr_pix) *                       (interpol->weight >> 24)
-					+*(ptr_pix + 1) *                   ((interpol->weight & 0xFFFFFF) >> 16)
-					+*(ptr_pix + priv->plugwidth) *     ((interpol->weight & 0xFFFF) >> 8)
-					+*(ptr_pix + priv->plugwidth + 1) * (interpol->weight & 0xFF)) >> 8;
+				const uint8_t *ptr_pix_end = priv->surface1 + (priv->plugwidth * priv->plugheight);
+
+				const uint8_t neighbor_right = (ptr_pix + 1 >= ptr_pix_end)
+						? 0
+						: *(ptr_pix + 1);
+				const uint8_t neighbor_below = (ptr_pix + priv->plugwidth >= ptr_pix_end)
+						? 0
+						: *(ptr_pix + priv->plugwidth);
+				const uint8_t neighbor_right_below = (ptr_pix + priv->plugwidth + 1 >= ptr_pix_end)
+						? 0
+						: *(ptr_pix + priv->plugwidth + 1);
+
+				color= (*(ptr_pix) * (interpol->weight >> 24)
+					+neighbor_right * ((interpol->weight & 0xFFFFFF) >> 16)
+					+neighbor_below * ((interpol->weight & 0xFFFF) >> 8)
+					+neighbor_right_below * (interpol->weight & 0xFF)) >> 8;
```

Comment:
I'm thinking this section could be written like this:

```c
/* Loop-invariant, bring it out of loop */
const uint8_t *ptr_pix_end = priv->surface1 + (priv->plugwidth * priv->plugheight);

...

const int y = interpol->coord & 0xffff;
const int x = interpol->coord >> 16;
ptr_pix = priv->surface1 + y * priv->plugwidth + x;

color = ptr_pix[0] * (interpol->weight >> 24);

/* right neighbour */
if (ptr_pix + 1 < ptr_pix_end) {
    color += ptr_pix[1] * ((interpol->weight >> 16) & 0xFF);
}

/* bottom neighbour */
if (ptr_pix + priv->plugwidth < ptr_pix_end) {
  color += ptr_pix[priv->plugwidth] * ((interpol->weight >> 8) & 0xFF);
}

/* bottom-right neighbour */
if (ptr_pix + priv->plugwidth + 1 < ptr_pix_end) {
  color += ptr_pix[priv->plugwidth + 1] * (interpol->weight & 0xFF);
}

color >>= 8;
```

Would be interesting to compare the generated, optimized code.

Reply:
@kaixiong I have pushed those^^ ideas as a dedicated commit on top now.

If you want to comare the assembly in [Godbolt's Compiler Explorer](https://gcc.godbolt.org), now you can:

# Before

https://gcc.godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1AB9U8lJL6yAngGVG6AMKpaAVxYM9DgDJ4GmADl3ACNMYhAANmkAB1QFQlsGZzcPPVj4mwFffyCWUPCoi0wrTIYhAiZiAmT3Ty4iksTyyoJswJCwyOkFCqqa1Pqelrbc/K6ASgtUV2Jkdg4AUgAmAGY/ZDcsAGoFlcce/EEAOgRd7AWNAEFltYYN123d/YJ8VBOzi%2BvV9c3MHb2DrQ8MF3itsFsAPQQrYsMS0VDIT6fHrEVzWLZUJgAa0wxmixDwADdjAQdgB2ABCny2Wz8pOibmAAHc8OgCKcVlSrjS6VsGa5gAhMHhBQRdlzLjTXHSABwkrYAKgUM0xcy44qRZIAIhqrsiCKj0ZiccY6WFYrR5QtKdSttLBCslvKmcLRbrrtr3Z9ebC/BBeZVgOYtsgEJVFQrA4TxuSJTSIITUKzxlGvdytii0aTjbj8UT5Qq0AwenyCYT/lqYXCERB4gAvTCoKi1g1ZjHY3Nlknjcbuml5wkAWjO/OZrPZFa2Sz7paJw7Bo6FIoQpN2lennNtA/n2GVxFVmC4k9htHhyAgXA0V97m/T27Oe4P6oArBSNAtn5W11sNKouOrb0lWchwfFUmDVD8KRfL8VkrX8lg3ONgJ3R9wMPSDp0/SdfxWFYZ3vMFUIg188Kw79f0kSQ0yAzMjQ7U1BHNFwCyLEszWIC0K1tGkaRPM9azwBsmxbQ1s3o9iLW7G8kIklwdxdZdV1gn8/1WSRnwiMkZVcaieUELYmHQdBjAUWYZ3tAguAieVVEnQNgwATzs4ggxnYhMGAPAejCO1ZQLaIDTxPBVBnXk0HhYhsN0yFoQpbAAHEAEkAltQzjNM5BJychVkJHRkWTZBAdiWCkthCwD%2B0C6JgsnAjdzAtVitKtKTLMwDbVY0kLLlUkFQC4ggtUYwHC49MeNywiGsPJqtggOrRwKiccvmxkl1FaSkXTTrfMEHqtn8ZdghIYwCVFUagPGuaqpq5ZSqPD5lP6wbhoYdAY12AAxH8thARUroG6rbNurYuA2raBBLbr5QOwUjoG0J4SZSd/sGmaVoFRaioeysnsBl63u48aieJknSfGz6f0JsnqZpnjfoVFHAbRssdwW8cEDBoDtqh0kYYQOGTsU4wEdQJG1yp2bcZukqJuwNnCpm%2B6wW/KWhocXsxpJin3014n6cZ6XSvRscFeB0HovCkhkYZ1WYxy/1GI4uSzgUs6HvBJZJHekqJaJvmBdOldFVmh2CCY2h5NdIPlgiFSPvjhP3uVsEQYib2kLJ/3jpFpH7dD8PI8U4rY9/BOPqT84U5ldPffGrOBsDghheKUXg/zp2I5dqPVyWEvVHjnt/kr85daJmVoqhLZsACHU7wJQRm2WJYhVPVAQxcK3lmfdAP0cBgl9IdeIs5213IIGYGEp9rPSuDhJloThn14TwOC0UhUE4RwM2mWY/huHhSAEE0HfSYWIQDPg0PoTgkhn7APfpwXgCgQCQKAa/O%2BpA4CwCQGgFg1V6BkAoBAHBeDOjIGAFwBCpAsCEjwHMAAangTATIADy0RGCcAATQWgYdiBIIgMEOBwQ/CVAchw3gQjmDEAcsw4I2hMDWDEaQHBbBBDMIYLQURaCqGYFhEYcQWj8DuWsESTASCtGYFUPI1wYdFFmgftwXgQJgjEBEc4LAcCDR4BYGIyYVADDAAUAwphrD2EOJkIIEQYh2BSHCfIJQag4G6HqAYIwIBTDGHME4pBkBJioACokMxiDijyNKPYV6/Q6ikB8H4doeROj1HSAkAQFS0hxCaQwEYHRwiDGKcYgQzQ%2BguFqHoSwJSmi9FaDU0Y9SLATJaYMCZnS6ndMmAoH%2BcwJD30frArRH8OBlRlBEQcUQthBkyhQo4SxZq4EIJvVYoNeCoK0D2UgYCIFQI4DA0gL8357MQcgwBwCXn2KWDs35CDAVoJeYSMI8Q7CSCAA%3D%3D%3D

```c
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h> // malloc

struct fake_priv_t {
  int plugwidth;
  int plugheight;
  uint8_t *surface1;
};

struct fake_interpol_t {
  uint32_t weight;
};

int main(int argc, char **argv) {
  (void)argv;

  struct fake_priv_t *const priv = malloc(sizeof(struct fake_priv_t));
  priv->plugwidth = 2;
  priv->plugheight = 2;
  priv->surface1 = malloc(1000);
  priv->surface1[0] = 0x11;
  priv->surface1[1] = 0x22;
  priv->surface1[2] = 0x33;
  priv->surface1[3] = 0x44;

  struct fake_interpol_t *const interpol =
      malloc(sizeof(struct fake_interpol_t));
  interpol->weight = 0x12345678u;

  int add_src;
  uint16_t x = argc, y = argc;
  register uint8_t *ptr_pix;
  int color = 0;

  // BEGIN
  add_src = y * priv->plugwidth + x;
  ptr_pix = priv->surface1 + add_src;

  const uint8_t *ptr_pix_end =
      priv->surface1 + (priv->plugwidth * priv->plugheight);

  const uint8_t neighbor_right =
      (ptr_pix + 1 >= ptr_pix_end) ? 0 : *(ptr_pix + 1);
  const uint8_t neighbor_below = (ptr_pix + priv->plugwidth >= ptr_pix_end)
                                     ? 0
                                     : *(ptr_pix + priv->plugwidth);
  const uint8_t neighbor_right_below =
      (ptr_pix + priv->plugwidth + 1 >= ptr_pix_end)
          ? 0
          : *(ptr_pix + priv->plugwidth + 1);

  color = (*(ptr_pix) * (interpol->weight >> 24) +
           neighbor_right * ((interpol->weight & 0xFFFFFF) >> 16) +
           neighbor_below * ((interpol->weight & 0xFFFF) >> 8) +
           neighbor_right_below * (interpol->weight & 0xFF)) >>
          8;

  // END
  printf("hello color %d\n", color);

  return 0;
}
```

# After

https://gcc.godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1AB9U8lJL6yAngGVG6AMKpaAVxYM9DgDJ4GmADl3ACNMYhAANmkAB1QFQlsGZzcPPVj4mwFffyCWUPCoi0wrTIYhAiZiAmT3Ty4iksTyyoJswJCwyOkFCqqa1Pqelrbc/K6ASgtUV2Jkdg4AUgAmAGY/ZDcsAGoFlcce/EEAOgRd7AWNAEFltYYN123d/YJ8VBOzi%2BvV9c3MHb2DrQ8MF3itsFsAPQQrYsMS0VDIT6fHrEVzWLZUJgAa0wxmixDwADdjAQdgB2ABCny2Wz8pOibmAAHc8OgCKcVlSrjS6VsGa5gAhMHhBQRdlzLjTXHSABwkrYAKgUM0xcy44qRZIAIhqrsiCKj0ZiccY6WFYrR5QtKdSttLBCslvKmcLRbrrtr3Z9ebC/BBeZVgOYtsgEJVFQrA4TxuSJTSIITUKzxlGvdytii0aTjbj8UT5Qq0AwenyCYT/lqYXCERB4gAvTCoKi1g1ZjHY3Nlknjcbuml5wkAWjO/OZrPZFa2Sz7paJw7Bo6FIoQpN2lennNtA/n2GVxFVmC4k9htHhyAgXA0V97m/T27Oe4P6oArBSNAtn5W11sNKouOrb0lWchwfFUmDVD8KRfL8VkrX8lg3ONgJ3R9wMPSDp0/SdfxWFYZ3vMFUIg188Kw79f0kSQ0yAzMjQ7U1BHNFwCyLEszWIC0K1tGkaRPM9azwBsmxbQ1s3o9iLW7G8kIklwdxdZdV1gn8/1WSRnwiMkZVcaieUELYmHQdBjAUWYZ3tAguAieVVEnQNgwATzs4ggxnYhMGAPAejCO1ZQLaIDTxPBVBnXk0HhYhsN0yFoQpbAAHEAEkAltBMk3QFMjJMszAP7QLomCycCN3MC1R2JYKS2JyFWQkdGRZNkEHKyqQsA21WNJCy5VJBUAuIILVGMBwuPTHjasI0rD2arYIGK0cGonGq5sZJdRWkpF03Ckgivy4LIPfLCav9RiOLks4FNFf5zjBKdJHWvV0yhLYCUu/wRWCEhbTwKgZr6gbpqPJ4%2BV2waHF7UaQxcbblipZS/oKkLX2gxUZuOggmNoeTXRXK6zi2KyY2WCIVIAMRJ%2B7rke6EPoIIgWC2N7gA%2B4gvp%2B2aQem5aBQWpqgfh4KhoYDLuMhiLmu/fnEYpLmx0aj9KyOtGMaxxTcZumVCaWYnfzJinbSemm6ee1XGeZ1nfo5mHxuwebx15ir8f%2BRxgf6hHBeFiGtsimGJZByCZZ5gH5ZRpXTsx87sdXLXSfJ6KvbV67KxlaKnuwAIdTvAlBGbZYliFU9UFF6GlmfdAP0cBhc9IIviD19N3IIGYGB/L1PSuDhJloThn14TwOC0UhUE4Z2FGmWY/huHhSAITQO8mLEQGfDR9E4SRe9nwfOF4BQQGXmf%2B470g4FgJA0BYAr6DICgIDPi/OmQYAuAQ0gsEJPA5gANTwTAmQAeWiRgnAp40FoOjYgO8IDBA3sEPwlQHJAN4DA5gxAHK/2CNoTA1gEGkDPmwQQv8GC0HgQfF%2BmBYRGHECQ/A7lrBEkwDvEhmBVCYNcOjbBZou7cF4ECYIxA4HOCwBvA0eAWAIMmFQAwwAFBfx/v/QBXCZCCBEGIdgUhFHyCUGoDeuh6gGCMCAUwxhzA8J3pASYqAAqJAYdvYomDSj2CFv0OopAfB%2BHaHkTo9R0gJAEE4tIcQfEMBGB0cIgxbG0IEM0PoLhah6EsHYpovRWhuNGJ4iwSS/GDCScEjxoTJijxmHMCQndu7rxIUPDgWxVAygiIOKIWwgzIHxksI4SwZq4EIMXFYXBxi8H3loHspAF5LxXhwNepA%2B4DwqdvXe09Z6DM4UsMpUyt5zIPoMwkYR4h2EkEAA

```c
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h> // malloc

struct fake_priv_t {
  int plugwidth;
  int plugheight;
  uint8_t *surface1;
};

struct fake_interpol_t {
  uint32_t weight;
};

int main(int argc, char **argv) {
  (void)argv;

  struct fake_priv_t *const priv = malloc(sizeof(struct fake_priv_t));
  priv->plugwidth = 2;
  priv->plugheight = 2;
  priv->surface1 = malloc(1000);
  priv->surface1[0] = 0x11;
  priv->surface1[1] = 0x22;
  priv->surface1[2] = 0x33;
  priv->surface1[3] = 0x44;

  struct fake_interpol_t *const interpol =
      malloc(sizeof(struct fake_interpol_t));
  interpol->weight = 0x12345678u;

  int add_src;
  uint16_t x = argc, y = argc;
  register uint8_t *ptr_pix;
  int color = 0;

  // BEGIN
  (void)add_src;
  ptr_pix = priv->surface1 + y * priv->plugwidth + x;

  const uint8_t *ptr_pix_end =
      priv->surface1 + (priv->plugwidth * priv->plugheight);

  color = ptr_pix[0] * (interpol->weight >> 24);

  // right neigbor
  if (ptr_pix + 1 < ptr_pix_end)
    color += ptr_pix[1] * ((interpol->weight >> 16) & 0xFF);

  // bottom neigbor
  if (ptr_pix + priv->plugwidth < ptr_pix_end)
    color += ptr_pix[priv->plugwidth] * ((interpol->weight >> 8) & 0xFF);

  // bottom right neigbor
  if (ptr_pix + priv->plugwidth + 1 < ptr_pix_end)
    color += ptr_pix[priv->plugwidth + 1] * (interpol->weight & 0xFF);

  color >>= 8;

  // END
  printf("hello color %d\n", color);

  return 0;
}
```

Reply:
@hartwork I have to say, the x64 assembly looked a bit alien to me at first with the new numbered registers and  a calling convention dramatically different from x32. It's been a while..

I made the functions closer to the Real Thing to avoid the compiler 'overfitting' its output and added `-O2 -fstrict-aliasing`. I also added a `test()` function just to clarify which register each argument to `compute_surface()` are stored in.

The outputs are here:
+ [Version 1](https://gcc.godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1AB9U8lJL6yAngGVG6AMKpaAVxYM9DgDJ4GmADl3ACNMYhAANmkAB1QFQlsGZzcPPVj4mwFffyCWUPCoi0wrTIYhAiZiAmT3Ty4iksTyyoJswJCwyOkFCqqa1Pqelrbc/K6ASgtUV2Jkdg4AUgAmAGY/ZDcsAGoFlcce/EEAOgRd7AWNAEFltYYN123d/YJ8VBOzi%2BvV9c3MHb2DrQ8MF3itsFsAPQQrYsMS0VDIT6fHrEVzWLZUJgAa0wxmixDwADdjAQdgB2ABCny2Wz8pOibmAAHc8OgCKcVlSrjS6VsGa5gAhMHhBQRdlzLjTXHSABwkrYAKgUM0xcy44qRZIAIhqrsiCKj0ZiccY6WFYrR5QtKdSttLBCslvKmcLRbrrtr3Z9ebC/BBeZVgOYtsgEJVFQrA4TxuSJTSIITUKzxlGvdytii0aTjbj8UT5Qq0AwenyCYT/lqYXCERB4gAvTCoKi1g1ZjHY3Nlknjcbuml5wkAWjO/OZrPZFa2Sz7paJw7Bo6FIoQpN2lennNtA/n2GVxFVmC4k9htHhyAgXA0V97m/T27Oe4P6oArBSNAtn5W11sNKouOrb0lWchwfFUmDVD8KRfL8VkrX8lg3ONgJ3R9wMPSDp0/SdfxWFYZ3vMFUIg188Kw79f0kSQ0yAzMjQ7U1BHNFwCyLEszWIC0K1tGkaRPM9azwBsmxbQ1s3o9iLW7G8kIklwdxdZdV1gn8/1WSRnwiMkZVcaieUELYmHQdBjAUWYZ3tAguAieVVEnQNgwATzs4ggxnYhMGAPAejCO1ZQLaIDTxPBVBnXk0HhYhsN0yFoQpbAAHEAEkAltQzjNM5BJychVkJHRkWTZBAdiWCkthCwD%2B0C6JgsnAjdzAtVitKtKTLMwDbVY0kLLlUkFQC4ggtUYwHC49MeNywiGsPJqtggOrRwKiccvmxkl1FaSkXTTrfMEHqtn8ZdghIYwCVFUagPGuaqpq5ZSqPD5lP6wbhoYdAY12AAxH8thARUroG6rbNurYuA2raBBLbr5QOwUjoG0J4SZSd/sGmaVoFRaioeysnsBl63u48aieJknSfGz6f0JsnqZpnjfoVFHAbRssdwW8cEDBoDtqh0kYYQOGTsU4wEdQJG1yp2bcZukqJuwNnCpm%2B6wW/KWhocXsxpJin3014n6cZ6XSvRscFeB0HovCkhkYZ1WYxy/1GI4uSzgUs6HvBJZJHekqJaJvmBdOldFVmh2CCY2h5NdIPlgiFSPvjhP3uVsEQYib2kLJ/3jpFpH7dD8PI8U4rY9/BOPqT84U5ldPffGrOBsDghheKUXg/zp2I5dqPVyWEvVHjnt/kr85daJmVoqhLZsACHU7wJQRm2WJYhVPVAQxcK3lmfdAP0cBgl9IdeIs5213IIGYGEp9rPSuDhJloThn14TwOC0UhUE4RwM2mWY/huHhSAEE0HfSYWIQDPg0PoTgkhn7APfpwXgCgQCQKAa/O%2BpA4CwCQGgFg1V6BkAoBAHBeDOjIGAFwBCpAsCEjwHMAAangTATIADy0RGCcAATQWgYdiBIIgMEOBwQ/CVAchw3gQjmDEAcsw4I2hMDWDEaQHBbBBDMIYLQURaCqGYFhEYcQWj8DuWsESTASCtGYFUPI1wYdFFmgftwXgQJgjEBEc4LAcCDR4BYGIyYVADDAAUAwphrD2EOJkIIEQYh2BSHCfIJQag4G6HqAYIwIBTDGHME4pBkBJioACokMxiDijyNKPYV6/Q6ikB8H4doeROj1HSAkAQFS0hxCaQwEYHRwiDGKcYgQzQ%2BguFqHoSwJSmi9FaDU0Y9SLATJaYMCZnS6ndMmAoH%2BcwJD30frArRH8OBlRlBEQcUQthBkyhQo4SxZq4EIJvVYoNeCoK0D2UgYCIFQI4DA0gL8357MQcgwBwCXn2KWDs35CDAVoJeYSMI8Q7CSCAA%3D%3D%3D)
+ [Version 2](https://gcc.godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1AB9U8lJL6yAngGVG6AMKpaAVxYM9DgDJ4GmADl3ACNMYhAANmkAB1QFQlsGZzcPPVj4mwFffyCWUPCoi0wrTIYhAiZiAmT3Ty4iksTyyoJswJCwyOkFCqqa1Pqelrbc/K6ASgtUV2Jkdg4AUgAmAGY/ZDcsAGoFlcce/EEAOgRd7AWNAEFltYYN123d/YJ8VBOzi%2BvV9c3MHb2DrQ8MF3itsFsAPQQrYsMS0VDIT6fHrEVzWLZUJgAa0wxmixDwADdjAQdgB2ABCny2Wz8pOibmAAHc8OgCKcVlSrjS6VsGa5gAhMHhBQRdlzLjTXHSABwkrYAKgUM0xcy44qRZIAIhqrsiCKj0ZiccY6WFYrR5QtKdSttLBCslvKmcLRbrrtr3Z9ebC/BBeZVgOYtsgEJVFQrA4TxuSJTSIITUKzxlGvdytii0aTjbj8UT5Qq0AwenyCYT/lqYXCERB4gAvTCoKi1g1ZjHY3Nlknjcbuml5wkAWjO/OZrPZFa2Sz7paJw7Bo6FIoQpN2lennNtA/n2GVxFVmC4k9htHhyAgXA0V97m/T27Oe4P6oArBSNAtn5W11sNKouOrb0lWchwfFUmDVD8KRfL8VkrX8lg3ONgJ3R9wMPSDp0/SdfxWFYZ3vMFUIg188Kw79f0kSQ0yAzMjQ7U1BHNFwCyLEszWIC0K1tGkaRPM9azwBsmxbQ1s3o9iLW7G8kIklwdxdZdV1gn8/1WSRnwiMkZVcaieUELYmHQdBjAUWYZ3tAguAieVVEnQNgwATzs4ggxnYhMGAPAejCO1ZQLaIDTxPBVBnXk0HhYhsN0yFoQpbAAHEAEkAltBMk3QFMjJMszAP7QLomCycCN3MC1R2JYKS2JyFWQkdGRZNkEHKyqQsA21WNJCy5VJBUAuIILVGMBwuPTHjasI0rD2arYIGK0cGonGq5sZJdRWkpF03Ckgivy4LIPfLCav9RiOLks4FNFf5zjBKdJHWvV0yhLYCUu/wRWCEhbTwKgZr6gbpqPJ4%2BV2waHF7UaQxcbblipZS/oKkLX2gxUZuOggmNoeTXRXK6zi2KyY2WCIVIAMRJ%2B7rke6EPoIIgWC2N7gA%2B4gvp%2B2aQem5aBQWpqgfh4KhoYDLuMhiLmu/fnEYpLmx0aj9KyOtGMaxxTcZumVCaWYnfzJinbSemm6ee1XGeZ1nfo5mHxuwebx15ir8f%2BRxgf6hHBeFiGtsimGJZByCZZ5gH5ZRpXTsx87sdXLXSfJ6KvbV67KxlaKnuwAIdTvAlBGbZYliFU9UFF6GlmfdAP0cBhc9IIviD19N3IIGYGB/L1PSuDhJloThn14TwOC0UhUE4Z2FGmWY/huHhSAITQO8mLEQGfDR9E4SRe9nwfOF4BQQGXmf%2B470g4FgJA0BYAr6DICgIDPi/OmQYAuAQ0gsEJPA5gANTwTAmQAeWiRgnAp40FoOjYgO8IDBA3sEPwlQHJAN4DA5gxAHK/2CNoTA1gEGkDPmwQQv8GC0HgQfF%2BmBYRGHECQ/A7lrBEkwDvEhmBVCYNcOjbBZou7cF4ECYIxA4HOCwBvA0eAWAIMmFQAwwAFBfx/v/QBXCZCCBEGIdgUhFHyCUGoDeuh6gGCMCAUwxhzA8J3pASYqAAqJAYdvYomDSj2CFv0OopAfB%2BHaHkTo9R0gJAEE4tIcQfEMBGB0cIgxbG0IEM0PoLhah6EsHYpovRWhuNGJ4iwSS/GDCScEjxoTJijxmHMCQndu7rxIUPDgWxVAygiIOKIWwgzIHxksI4SwZq4EIMXFYXBxi8H3loHspAF5LxXhwNepA%2B4DwqdvXe09Z6DM4UsMpUyt5zIPoMwkYR4h2EkEAA)

It's hard to tell which is faster on modern CPUs. So I think I'm giving up and will go run some benchmarks instead..

Reply:
@hartwork, I modified the code above to include the rendering loop to prevent the GCC optimizer from overfitting to the test and then added the flags `-O3 -fstrict-aliasing`. After spending _way too much_ time on Godbolt studying the x64 assembly (it's been a while!), I concluded it was impossible to tell which looked faster :laughing:

GCC's optimizer interleaves register use to keep the pipeline as full as possible, making it harder to read or estimate the clock count. Apart from that, the assembly code was pretty pedestrian i.e. had nothing fancy like conditional moves, etc.

So I ran `actor_benchmark` (release build with `-O3 -fstrict-aliasing`) on the two versions and confirmed that my rewritten version is about 10% faster (at 640x400):

Version 1
```
$ ./actor_bench 10000 infinite
-- ActorBench --
Total runs: 10000
Total time: 7.80415e+06us
Time / run: 780.415us

$ ./actor_bench 10000 infinite
-- ActorBench --
Total runs: 10000
Total time: 7.85247e+06us
Time / run: 785.247us
```

Version 2
```
$ ./actor_bench 10000 infinite
-- ActorBench --
Total runs: 10000
Total time: 6.92998e+06us
Time / run: 692.998us

$ ./actor_bench 10000 infinite
-- ActorBench --
Total runs: 10000
Total time: 7.08624e+06us
Time / run: 708.624us
```

Time/Run is the average frame time.

So let's go with version 2. It's also a bit nicer to read.

Reply:
@kaixiong cool!